src/runmode-dpdk.c

```diff
@@ -0,0 +1,1295 @@
+        SCReturnPtr(NULL, "char *");
+    }
+
+    SCReturnPtr(ptr, "char *");
+}
+
+/**
+ * Allocates space for length of the given string and then copies contents
+ * @param arg String to set to the newly allocated space
+ * @return memory address if no error otherwise NULL (with errno set)
+ */
+char *AllocAndSetArgument(const char *arg)
+{
+    if (arg == NULL) {
+        SCLogError(SC_ERR_DPDK_CONF, "Passed argument is NULL in DPDK config initialization");
+        errno = EINVAL;
+        SCReturnPtr(NULL, "char *");
+    }
+
+    char *ptr;
+    size_t arg_len = strlen(arg);
+
+    ptr = AllocArgument(arg_len);
+    if (ptr == NULL)
+        SCReturnPtr(NULL, "char *");
+
+    strlcpy(ptr, arg, arg_len + 1);
+    SCReturnPtr(ptr, "char *");
+}
+
+char *AllocAndSetOption(const char *arg)
+{
+    if (arg == NULL) {
+        errno = EINVAL;
+        SCReturnPtr(NULL, "char *");
+    }
+
+    char *ptr = NULL;
+    size_t arg_len = strlen(arg);
+    uint8_t is_long_arg = arg_len > 1;
+    const char *dash_prefix = is_long_arg ? "--" : "-";
+    size_t full_len = arg_len + strlen(dash_prefix);
+
+    ptr = AllocArgument(full_len);
+    if (ptr == NULL)
+        SCReturnPtr(NULL, "char *");
+
+    strlcpy(ptr, dash_prefix, strlen(dash_prefix) + 1);
+    strlcat(ptr, arg, full_len + 1);
+    SCReturnPtr(ptr, "char *");
+}
+
+int ArgumentsInit(struct Arguments *args, unsigned capacity)
+{
+    args->argv = SCCalloc(capacity, sizeof(args->argv));
+    if (args->argv == NULL) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Could not allocate memory for Arguments structure");
+        SCReturnInt(-ENOMEM);
+    }
+
+    args->capacity = capacity;
+    args->argc = 0;
+    SCReturnInt(0);
+}
+
+void ArgumentsCleanup(struct Arguments *args)
+{
+    for (int i = 0; i < args->argc; i++) {
+        if (args->argv[i] != NULL) {
+            SCFree(args->argv[i]);
+            args->argv[i] = NULL;
+        }
+    }
+
+    SCFree(args->argv);
+    args->argv = NULL;
+    args->argc = 0;
+    args->capacity = 0;
+}
+
+int ArgumentsAdd(struct Arguments *args, char *value)
+{
+    if (args->argc + 1 > args->capacity) {
+        SCLogError(SC_ERR_DPDK_EAL_INIT, "No capacity for more arguments");
+        SCReturnInt(-ENOBUFS);
+    }
+
+    args->argv[args->argc++] = value;
+    SCReturnInt(0);
+}
+
+int ArgumentsAddOptionAndArgument(struct Arguments *args, const char *opt, const char *arg)
+{
+    int retval;
+    char *option;
+    char *argument;
+
+    option = AllocAndSetOption(opt);
+    if (option == NULL)
+        SCReturnInt(-errno);
+
+    retval = ArgumentsAdd(args, option);
+    if (retval < 0) {
+        SCFree(option);
+        SCReturnInt(retval);
+    }
+
+    // Empty argument could mean option only (e.g. --no-huge)
+    if (arg == NULL || arg[0] == '\0')
+        SCReturnInt(0);
+
+    argument = AllocAndSetArgument(arg);
+    if (argument == NULL) {
+        SCReturnInt(-errno);
+    }
+    ArgumentsAdd(args, argument);
+    SCReturnInt(0);
```

Comment:
I don't see any leaks in L232. Address of the newly allocated option is already added to `args` structure and because the `arg` does not exist/is empty, nothing is allocated. I am only freeing the `option` variable on L226 because from the return value I assume it was not possible to insert into the `args` structure.

The same as previously described applies for L236 since it could return NULL only when it fails to allocate space for the `argument`. The option is already in `args` structure. However, if FatalErrors will be in place (instead of returned error codes) then L236 and L238 will not require error checks.